apps/lib/publisher_agent/metadata.go

```diff
@@ -0,0 +1,153 @@
+package publisher_agent
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"runtime"
+	"time"
+
+	"github.com/Stork-Oracle/stork-external/apps/lib/signer"
+	"github.com/rs/zerolog"
+)
+
+const publicIpUrl = "https://api.ipify.org"
+const awsMetadataUrl = "http://169.254.169.254/latest/meta-data"
+const versionFile = "version.txt"
+
+type PublisherMetadata struct {
+	PublisherKey          signer.PublisherKey  `json:"publisher_key"`
+	SignatureType         signer.SignatureType `json:"signature_type"`
+	PublisherAgentVersion string               `json:"publisher_agent_version"`
+	Architecture          string               `json:"architecture"`
+	PublicIp              string               `json:"public_ip"`
+	AwsMetadata           AwsMetadata          `json:"aws_metadata"`
+}
+
+type AwsMetadata struct {
+	IsAws              bool   `json:"is_aws"`
+	InstanceType       string `json:"instance_type"`
+	AvailabilityZoneId string `json:"availability_zone_id"`
+}
+
+type PublisherMetadataReporter struct {
+	publicKey           signer.PublisherKey
+	signatureType       signer.SignatureType
+	reportPeriod        time.Duration
+	storkRestApiBaseUrl string
+	storkAuth           AuthToken
+	logger              zerolog.Logger
+}
+
+func NewPublisherMetadataReporter(
+	publicKey signer.PublisherKey,
+	signatureType signer.SignatureType,
+	reportPeriod time.Duration,
+	storkRestApiBaseUrl string,
+	storkAuth AuthToken,
+	logger zerolog.Logger,
+) *PublisherMetadataReporter {
+	return &PublisherMetadataReporter{
+		publicKey:           publicKey,
+		signatureType:       signatureType,
+		reportPeriod:        reportPeriod,
+		storkRestApiBaseUrl: storkRestApiBaseUrl,
+		storkAuth:           storkAuth,
+		logger:              logger,
+	}
+}
+
+func (p *PublisherMetadataReporter) Run() {
+	err := p.report()
+	if err != nil {
+		p.logger.Warn().Msgf("Error reporting publisher metadata: %v", err)
+	}
+	for range time.Tick(p.reportPeriod) {
```

Comment:
it's maybe a little silly to push this metadata more than once since nothing is expected to change.

It's nice to be able to treat it as a heartbeat too though - if a particular key hasn't reported for an hour then we can assume the agent has stopped. Also useful if we add more dynamic metadata in the future.